Add command to get a new cluster service account

charlieegan3 opened the pull request:
Fixes #63

```
$ go run main.go clusters connect charlie --service-account-only
apiVersion: v1
data:
  credentials.json: xxx
kind: Secret
metadata:
  creationTimestamp: null
  name: agent-credentials
  namespace: jetstack-secure
type: Opaque
$ go run main.go clusters connect charlie --service-account-only --service-account-format=json
{
  "user_id": "xxx",
  "user_secret": "xxx"
}
```

Signed-off-by: Charlie Egan <charlieegan3@users.noreply.github.com>

hawksight reviewed (approved):
I can't speak for the code, but the outcome looks to be exactly what I was looking for, thank you :+1:

irbekrm reviewed (changes requested):
I had another look at the jsctl commands that we already have.

As I understand it would now already be possible to achieve #63  if the existing global `--stdout` flag had an option to specify format, you could just
```
jsctl clusters connect --stdout --stdout-format=json
```
Additionally we already have one command that just outputs a credential (in this case for JSE GCR and again has differently named flags
```
jsctl registry auth output --format=json
```

Conscious that we might be adding these small flags all over the place that might be difficult to find and also each have different naming.

What would you think about having a single `jsctl auth` command that could have subcommands to output either JSE GCR credential or agent token in the specified format? I think this would make more sense from user perspective as in cases where they just need creds the `cluster connect` etc commands are otherwise irrelevant for them.

Perhaps for now, if we don't want to implement this new command we could just add another global flag `--stdout-json` (which probably would sooner or later be useful anyway) and the agent SA could just be retrieved by running
```
jsctl clusters connect --stdout --stdout-format=json | jq -r .<path-to-agent-secret>
```

wdyt @charlieegan3 @hawksight ?

charlieegan3 commented:
> jsctl clusters connect --stdout --stdout-format=json

Yeah, we could do this. However the user would need to unbase64 decode the value from the secret. I was under the impression that just getting the service account json would be most useful.

> jsctl registry auth output --format=json

You are correct, the command flag format for this is different. One difference with this PR is that the credentials are created for each new cluster to identify them, rather than the registry creds which can be reused.

> What would you think about having a single jsctl auth command that could have subcommands to output either JSE GCR credential or agent token in the specified format? 

I think that this might be nicer than getting users to use --stdout and overloading the connect command as my PR does. This way we can copy the flags of the `auth output` command more closely (however, a cluster name will be required).

I can update this PR to do that, it's not a lot of work.

charlieegan3 commented:
@irbekrm I have updated this now with a new command under auth. I think this is a much better place for it. Wdyt?

irbekrm commented on internal/command/auth/clusters.go:
```diff
@@ -0,0 +1,89 @@
+
+	cmd := &cobra.Command{
+		Use:   "create-service-account [name]",
+		Short: "Create a new service account identity for a cluster",
```
Suggesion: `generate JSS agent credentials for a cluster`

charlieegan3 replied:
I have updated this in 0acc3bb068bef8b517564899df5c394a5da0264e, but again used a different phrasing.

irbekrm commented on internal/command/auth/clusters.go:
```diff
@@ -0,0 +1,89 @@
+				secret := cluster.AgentServiceAccountSecret(serviceAccountBytes)
+				secretYAMLBytes, err := yaml.Marshal(secret)
+				if err != nil {
+					return fmt.Errorf("failed to marshal image pull secret: %s", err)
```
`failed to marshal agent credentials secret`

charlieegan3 replied:
03fd65da545437114921969ea03e314431f64eac - thanks!

irbekrm commented on internal/cluster/cluster.go:
```diff
@@ -158,3 +161,24 @@ func marshalBase64(in interface{}) ([]byte, error) {
 	return buffer.Bytes(), nil
 }
+
+// AgentServiceAccount secret takes a service account json and formats it as a
```
Question: do we use 'service account' naming here and elsewhere because the agent auth is based on GCP service accounts? The returned credential does not look like a service account value to me. Should we rather call it 'agent credential' or something such to avoid confusion?

charlieegan3 replied:
I have updated the command text to make this clearer, I hope. I haven't renamed the command as I think that the command name is still accurate.

irbekrm commented on internal/command/auth/clusters.go:
```diff
@@ -0,0 +1,89 @@
+		Use:   "create-service-account [name]",
+		Short: "Create a new service account identity for a cluster",
+		Args:  cobra.MatchAll(cobra.ExactArgs(1)),
+		Long: `jsctl can do this automatically for you, in jsctl clusters connect. However,
```
Suggestion: `generate JSS agent credentials for authentication against JSS UI. This is only needed if you are not deploying the agent with jsctl. Output can be json formatted credential or a Kubernetes Secret with the credential for `jetstck-secure` namespace.`

(given that we want to eventually deploy the agent with the operator, we probably don't want to talk about doing it via `jsctl clusters connect` too much..)

charlieegan3 replied:
That's fair enough. I have updated this in 0deda09281664dee71ecea1a74e3aae3ec0bcf8f. However, have adjusted the wording a little.

irbekrm commented on internal/cluster/cluster.go:
```diff
@@ -158,3 +161,24 @@ func marshalBase64(in interface{}) ([]byte, error) {
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "agent-credentials",
+			Namespace: "jetstack-secure",
```
I think the namespace will probably need to be modifyable. However, perhaps that can be done in a different PR when/if that gets asked for

irbekrm reviewed (changes requested):
Thanks for working on this Charlie! I found the service account naming confusing and left a bunch of suggestions for that.
Additionally it looks like `--stdout` flag for `clusters connect` command is somehow broken in this PR, when I run `go run main.go cluster connect foo --stdout` I get manifests applied to cluster instead of output to stdout, not sure if you see that as well?

charlieegan3 commented on internal/command/auth/clusters.go:
```diff
@@ -0,0 +1,89 @@
+	flags.StringVar(
+		&serviceAccountFormat,
+		"format",
+		"json",
```
I wondered about this too. Wasn't sure what to call it but jsonKeyData is as good as any and it's clearer than json vs secret as you say.

Fixed in d998f52a6f14703dd3cabc6997ab4b6101978dde

charlieegan3 commented on internal/command/auth/clusters.go:
```diff
@@ -0,0 +1,89 @@
+	var serviceAccountFormat string
+
+	cmd := &cobra.Command{
+		Use:   "create-service-account [name]",
```
See https://github.com/jetstack/jsctl/pull/67#discussion_r1026555067

These are service accounts, just for Jetstack secure rather than GCP or K8s.

I have updated the command text to 'Create a new Jetstack Secure service account for a cluster agent' in another commit. I'm happy that this addresses that confusion enough for now.

irbekrm reviewed (approved):
Thank you for making the changes @charlieegan3 this seems clearer to me.

I guess it would be good if we used the terminology consistently and I don't see service accounts (or any reference to creds) in JSS agent docs, however that's not to do with this PR.

One minor nit is that the help output still suggests to me that 'name' argument is optional which it is not:
```
go run main.go auth clusters create-service-account --help
...
Usage:
  jsctl auth clusters create-service-account [name] [flags]
```

But happy for this to be merged either way, great work!

/lgtm

charlieegan3 commented:
> I guess it would be good if we used the terminology consistently

Hmm. Yeah, that would be good. This does make the CLI consistent with what is actually happening in the API. We can pick that up in a review of the docs. It might be that this is all adjusted anyway if the agent is to be deployed with the operator in the near future.

> suggests to me that 'name' argument is optional which it is not

Hmm, that's a good point. I've updated this command and various other commands in the application that used the same unclear formatting so it's consistent.

charlieegan3 commented:
Thanks Irbe. I've rebased this, updated the docs and merged it in.

@hawksight, this will be in the next release but you can build from source here if you need it sooner!
